src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -319,7 +319,7 @@ private string ConcatAsString<T>(IEnumerable<T> enumerable)
             {
                 cacheFile = FileUtility.SafeRead(_request.Project.RestoreMetadata.CacheFilePath, (stream, path) => CacheFileFormat.Read(stream, _logger, path));
 
-                if (cacheFile.IsValid && StringComparer.Ordinal.Equals(cacheFile.DgSpecHash, newDgSpecHash))
+                if (cacheFile.IsValid && StringComparer.Ordinal.Equals(cacheFile.DgSpecHash, newDgSpecHash) && _request.ExistingLockFile.PackageSpec.FilePath.Equals(_request.Project.FilePath))
```

Comment:
you need to get the file system based string comparer from PathUtility.GetComparerBasedOnOS

Reply:
and add a test ;)

Reply:
For test hints, look at https://github.com/NuGet/NuGet.Client/blob/cdf58815319ec0c9d1a29c3c790233826c1bab8a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs. 

There is infrastructure for project creation etc. 
You can also just create one project and then create a copy with a different name. 

You should be able to test the exact scenario described in the issue.